disnake/enums.py

```diff
@@ -72,324 +85,1357 @@
 )
 
 
-class _EnumValueBase(NamedTuple):
-    if TYPE_CHECKING:
-        _cls_name: ClassVar[str]
+_T = TypeVar("_T", bound="EnumMeta")
 
-    name: str
-    value: Any
 
-    def __repr__(self) -> str:
-        return f"<{self._cls_name}.{self.name}: {self.value!r}>"
+def _is_descriptor(obj):
+    return hasattr(obj, "__get__") or hasattr(obj, "__set__") or hasattr(obj, "__delete__")
 
-    def __str__(self) -> str:
-        return f"{self._cls_name}.{self.name}"
 
+class _EnumDict(Dict[str, Any]):
+    def __init__(self, base: Type[Any]):
+        # We explicitly take a base in __init__, unlike default Enums, to more easily
+        # enforce proper value typing. This should ensure better performance, as e.g.
+        # `Enum[int].one + 3 == 4` is more performant than `Enum[Any].one.value + 3 == 4`,
+        # as this saves a lookup. We do lose some functionality here, but it's probably
+        # worth the cost; especially considering all current enums have monotyped values.
+        super().__init__()
+        self.base = base
+        self.member_map: Dict[str, Any] = {}
+        self.value_map: Dict[Any, str] = {}
 
-@total_ordering
-class _EnumValueComparable(_EnumValueBase):
-    def __eq__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value == other.value
+    def __getitem__(self, name: str) -> Any:
+        try:
+            return super().__getitem__(name)
+        except KeyError:
+            return self.value_map[name]
```

Comment:
This doesn't make sense to me. `value_map` is a mapping of member values -> member names, and `__getitem__` takes in a `name` parameter.

Reply:
Correct, that should've been `self.member_map[name]`. Either way, after some testing, this entire method seems to no longer be necessary, so it has been removed in fce05efc82105113ab5bb08463c74594b5cdf3d2.